api/src/services/purged-docs.js

```diff
@@ -51,59 +65,109 @@ const getPurgedIdsFromChanges = result => {
+};
+
+const getPurgedIds = async (userCtx, docIds, useCache = true) => {
   let purgeIds = [];
-  let cache;
-  let cacheKey;
+  const cacheDocId = getCacheDocId(userCtx.name);
   if (!docIds?.length || !userCtx.roles?.length) {
     return Promise.resolve(purgeIds);
   }
 
   if (useCache) {
-    cache = cacheService.instance(CACHE_NAME);
-    cacheKey = getCacheKey(userCtx, docIds);
-    const cached = cache.get(cacheKey);
-    if (cached) {
-      cache.ttl(cacheKey);
-      return Promise.resolve(cached);
+    const cachedIds = await getCachedIds(cacheDocId);
+    if (cachedIds) {
+      return Promise.resolve(cachedIds);
     }
   }
 
   const ids = docIds.map(purgingUtils.getPurgedId);
   let purgeDb;
-  // requesting _changes instead of _all_docs because it's roughly twice faster
-  return getPurgeDb(userCtx.roles)
-    .then(tempDb => purgeDb = tempDb)
-    .then(() => purgeDb.changes({ doc_ids: ids, batch_size: ids.length + 1, seq_interval: ids.length }))
-    .then(result => {
-      purgeIds = getPurgedIdsFromChanges(result);
-      cache?.set(cacheKey, purgeIds);
-      db.close(purgeDb);
-    })
-    .catch(err => catchDbNotFoundError(err, purgeDb))
-    .then(() => purgeIds);
+
+  try {
+    purgeDb = await getPurgeDb(userCtx.roles);
+    // requesting _changes instead of _all_docs because it's roughly twice faster
+    const changesResult = await purgeDb.changes({ doc_ids: ids, batch_size: ids.length + 1, seq_interval: ids.length });
+    purgeIds = getPurgedIdsFromChanges(changesResult);
+    useCache && await setCachedIds(cacheDocId, purgeIds);
+    db.close(purgeDb);
+  } catch (err) {
+    catchDbNotFoundError(err, purgeDb);
+  }
+
+  return purgeIds;
 };
 
 const getUnPurgedIds = (userCtx, docIds) => {
+  docIds.sort();
   return getPurgedIds(userCtx, docIds).then(purgedIds => _.difference(docIds, purgedIds));
 };
 
-const listen = (seq = 'now') => {
-  db.sentinel
-    .changes({ live: true, since: seq })
-    .on('change', change => {
-      seq = change.seq;
-      if (change.id.startsWith('purgelog:') && change.changes[0].rev.startsWith('1-')) {
-        clearCache();
-      }
-    })
-    .on('error', err => {
-      logger.error('Error watching sentinel changes, restarting: %o', err);
-      listen(seq);
-    });
+const listen = () => {
+  watchSentinel();
+  watchUsers();
+  watchMedic();
+  watchSettings();
+};
+
+const watchSentinel = () => {
+  dbWatcher.sentinel(change => {
+    if (change.id.startsWith('purgelog:') && change.changes[0].rev.startsWith('1-')) {
+      return clearCache();
+    }
+  });
+};
+
+const watchUsers = () => {
+  dbWatcher.users(({ id }) => {
+    if (id.startsWith(USER_DOC_PREFIX)) {
+      const name = id.replace(USER_DOC_PREFIX, '');
+      return clearCache(name);
+    }
+  });
+};
+
+const watchMedic = () => {
+  dbWatcher.medic(({ id }) => {
+    if (id.startsWith(USER_DOC_PREFIX)) {
+      const name = id.replace(USER_DOC_PREFIX, '');
+      return clearCache(name);
+    }
+  });
+};
```

Comment:
Why do we need this as well as watching _users?

Reply:
Because we store `contact_id` in the `user-settings` document and `facility_id` in the `user`  document :( 
https://github.com/medic/cht-core/blob/a686679a6a4d197c40c0036c25733b4ff96716aa/shared-libs/user-management/src/users.js#L761

Reply:
Ahh ok! Thanks.